app/controllers/application_controller.rb

```diff
@@ -5,6 +5,139 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, 'this_is_a_secret'
   end
 
+  helpers do 
+    def current_user
+      User.find(session[:user_id])
+     end
+ 
+     def logged_in?
+       !!session[:user_id]
+     end
+  end
+
+  get '/' do 
+    erb :'/index'
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect '/tweets'
+    else
+    erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do 
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect '/signup'
+    else
+    @user = User.create(params)
+    session[:user_id] = @user.id 
+    redirect '/tweets'
+    end
+  end
+
+  get '/login' do 
+    if logged_in?
+      redirect '/tweets'
+    else
+    erb :'users/login'
+    end
+  end
+
+  post '/login' do 
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+    session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect 'users/login'
+    end
+  end
+
+  get '/tweets' do 
+    if logged_in?
+    @tweets = Tweet.all
+    erb :'tweets/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/new' do 
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do 
+    if params[:content].empty?
+      redirect '/tweets/new'
+    else
+      @tweet = Tweet.create(content: params[:content], user: current_user)
+      redirect '/tweets'
+    end
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do 
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if current_user.id == @tweet.user_id
+        erb :'tweets/edit_tweet'
+      else
+        redirect '/tweets'
+      end
+    else 
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/:id/edit' do 
+    if params[:content] == ""
+      redirect "/tweets/#{@tweet.id}/edit"
+    else
+      @tweet = Tweet.find_by_id(params[:id])
+      @tweet.content = params[:content]
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    end
+
+  end
+
+  post '/tweets/:id/delete' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if current_user.id == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.